File: 8. String to Integer (atoi) Medium/solution.go
package main

import (
	"fmt"
	"math"
	"strings"
)

func main() {
	fmt.Println(myAtoi("42"))
	fmt.Println(myAtoi(" -042"))
	fmt.Println(myAtoi("1337c0d3"))
	fmt.Println(myAtoi("0-1"))
	fmt.Println(myAtoi("words and 987"))
}

func myAtoi(s string) int {
	s = strings.TrimSpace(s)

	if len(s) == 0 {
		return 0
	}

	sign := 1
	if s[0] == '-' {
		sign = -1
		s = s[1:]
	} else if s[0] == '+' {
		s = s[1:]
	}

	var result int
	for _, char := range s {
		if char < '0' || char > '9' {
			break
		}
		digit := int(char - '0')
		result = result*10 + digit

		if result*sign > math.MaxInt32 {
			return math.MaxInt32
		} else if result*sign < math.MinInt32 {
			return math.MinInt32
		}
	}

	return result * sign
}

File: 8. String to Integer (atoi) Medium/solution.java
class solution {
    public static void main(String[] args) {
        System.out.println(myAtoi("42"));
        System.out.println(myAtoi(" -042"));
        System.out.println(myAtoi("1337c0d3"));
        System.out.println(myAtoi("0-1"));
        System.out.println(myAtoi("words and 987"));
    }

    public static int myAtoi(String s) {
        int index = 0, sign = 1, result = 0;
        while (index < s.length() && s.charAt(index) == ' ')
            index++;
        if (index < s.length() && (s.charAt(index) == '-' || s.charAt(index) == '+')) {
            sign = (s.charAt(index) == '-') ? -1 : 1;
            index++;
        }
        while (index < s.length() && Character.isDigit(s.charAt(index))) {
            int digit = s.charAt(index) - '0';
            if (result > Integer.MAX_VALUE / 10 || (result == Integer.MAX_VALUE / 10 && digit > 7)) {
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            result = result * 10 + digit;
            index++;
        }
        return result * sign;
    }
}
